### ISSUE TITLE
Create reports as part of automatic post-upload workflow

### ISSUE THREAD

**Issue by dpark01** [dnanexus/viral-ngs#25]
By popular demand among our researchers: it's highly useful to see the number of reads that align to ERCC spike-ins for each raw demultiplexed bam file. It's also computationally cheap enough that it doesn't matter if we run it even when it's not applicable.

Can you spin off jobs, post-demultiplexing (pre-depletion) that runs this:
`read_utils.py align_and_count_hits samplename.raw.bam spikeins_db.fasta samplename.spike_count.txt`
You can see an example of this invocation in pipes/rules/reports.rules

If you don't have the spikeins_db.fasta I can email it to you, it's quite tiny (32 sequences, 1kb each).

Additionally, perhaps spin off your fastqc applet on the raw bams as well? (project-B406G0x2fz2B3GVk65200003:applet-BgY1x2Q087VvzyzQG1f68y0v)

**Comment by dpark01** [comment]
Here's the file: http://www.broadinstitute.org/~dpark/ercc_spike-ins.fasta

**Comment by yifei-men** [comment]
Hi Danny,

Got it, thanks for the writeup.

Re-iterating your request to make sure I understand them. What you're asking is, in essence, 

```
for bam_file in {demux_output}; do
    read_utils.py align_and_count_hits $bam_file ....
    [upload samplename.spike_count.txt as an output]

    fastqc $bam_file 
    [upload fastqc reports as output]
```

### Implementation

For logistics and performance reasons, my intuition is to just tag the `read_utils.py` command into the current `demux` applet (so files don't need to be pulled down then staged again).

We'll run fastqc as a separate step in the workflow since it's already an existing app and that will just make things easier.

Does that sound ok from your end?

**Comment by dpark01** [comment]
Sounds exactly right on the essence

My only thought on implementation is that maybe you could throw it into metagenomics instead of demux? Because I can imagine wanting that report on BAMs that were not generated by demux. Or bundled into deplete, maybe that makes less sense though. Metagenomics makes sense because that also generates numeric reports based on read content. Only problem though is that putting it in metagenomics doesn't parallelize as well as putting it in deplete..

I'm open to what your preferences are on implementation.

**Comment by yifei-men** [comment]
> Because I can imagine wanting that report on BAMs that were not generated by demux.

I see, then I think I'll just pull out the read_utils step as a standalone app, give us flexibility... 

I also don't think that'll really affect cost / time too significantly considering deplete and metagenomics will be a lot more expensive / time-consuming anyways..

**Comment by yifei-men** [comment]
@dpark01 @tomkinsc 

I've done a trial of `read_utils.py align_and_count_hits` as a downstream step of demux (sample run [here](https://platform.dnanexus.com/projects/BXBXK180x0z7x5kxq11p886f/monitor/analysis/BxykpQ80x0z9PqbQ39jYj1yp)

What was immediately striking was:
The run time of `align_and_count_hits` is actually quite long... it's obviously quite sensitive to the size of the demux-ed bam, but in this MiSeq run, the largest bam took more than 3 hours to align and count, which seem a bit prohibitive, since the demux/metagenomics pipeline only took ~ 1/2 hour. The majority is taken up by novoalign. Does this runtime make sense at all? (Caveat, this is run single-threaded in the novoalign step, see `#1` below)

There are some things that came to mind for me, but any suggestions from your end are definitely appreciated:
1. This is run without a Novoalign license file (the changes to support #26 is already in place). Could you try to relaunch this [job](https://platform.dnanexus.com/projects/BXBXK180x0z7x5kxq11p886f/monitor/job/BxykpQ80x0z9PqbQ39jYj1yy) and supply a Novoalign license file to see if multithreading actually happen? (I couldn't find a license in our original novocraft tarball 😞 )... 
2.  Is there a parameter (something like subsampling?) that we can pass to Novoalign to reduce the alignment step runtime... although I guess that defeats the purpose of counting hits...

**Comment by dpark01** [comment]
Hm, okay that's surprising, but perhaps that's just the way it is and we never really characterized it before. A few thoughts...
1. While we could run this with a Novoalign license file to support threading, we haven't yet done anything in the automated demux workflow that requires the Novoalign license or GATK or some sort of user-supplied additional input. If we did allow for the license file to be specified, I'm wondering how we'd do that in the automated workflow executions..
2. Novoalign does have a few parameters that can change runtime vs sensitivity, but I kind of thought we were already running it without the extra-sensitive parameters that refine1 uses.
3. Because these are spike-in sequences and not some sort of highly diverse virus, I'm leaning towards thinking that maybe we should just use BWA MEM instead, since there's really no reason to need to use an aligner so highly tuned for sensitivity.

@yifei-men if we wanted to try out an alternate version that uses BWA, would you want to give that a try yourself (it could be your standard BWA applet, maybe followed by samtools idxstats, and we could just use that file as output). Or would you prefer us to implement it in viral-ngs read_utils.py and pass it to you that way?

**Comment by yifei-men** [comment]
@dpark01 

> If we did allow for the license file to be specified, I'm wondering how we'd do that in the automated workflow executions.

[dx-streaming-upload](https://github.com/dnanexus-rnd/dx-streaming-upload#example-playbook) has a recent change that accepts downstream inputs specified as a JSON file (see the field `downstream_input` in the example playbook)... I was gonna email you with a list of the recent changes when we squash the issues here so you can do an epic update instead of drib and drab updates, so sorry for the gap... I think the idea would be to chain in the license file (stored in a private-ish project that the job launcher can access) in automated workflows.

> if we wanted to try out an alternate version that uses BWA, would you want to give that a try yourself (it could be your standard BWA applet, maybe followed by samtools idxstats, and we could just use that file as output).

I can just put together a quick and dirty implementation using our own BWA and samtools and see what that churns out... I'll ping you when I get something out (probably by this week) and we can take a look together to see if that's suitable enough or you'd prefer an in-house wrapper.

**Comment by yifei-men** [comment]
Hey @dpark01 

This is a sample [run](https://platform.dnanexus.com/projects/BXBXK180x0z7x5kxq11p886f/monitor/job/BxzGQVj0x0zKZ7yf3Jyzkkjg) with a drop-in replacement of `bwa-mem + samtools idxstats`. 

The FastQC is still there as previously. The run time is definitely a lot better (a couple of minutes)

Does this output look reasonable based on the asks you received from users?

**Comment by yifei-men** [comment]
Addressed by #30

@dpark01, let me know if the `bwa-samtools` output serves your needs. 

I'll leave this open for you to close as you see appropriate.
